README.md
Language: [English](README.md) | [中文简体](README-ZH.md)

<p align="center">
  <img width="192" src="https://img-blog.csdnimg.cn/20201214142404283.png" alt="axios mapper">
</p>

<p align="center">
  <a href="https://www.npmjs.com/package/axios-mapper">
    <img src="https://img.shields.io/npm/v/axios-mapper?color=blue" alt="npm">
  </a>
   <a href="https://www.npmjs.com/package/axios-mapper">
    <img src="https://img.shields.io/node/v/axios-mapper" alt="npm">
  </a>
  <a href="https://www.npmjs.com/package/axios-mapper">
    <img src="https://img.shields.io/npm/dw/axios-mapper" alt="npm">
  </a>
   <a href="https://www.npmjs.com/package/axios-mapper">
    <img src="https://img.shields.io/bundlephobia/minzip/axios-mapper" alt="npm">
  </a>
  <a href="https://github.com/RainManGO/axios-mapper/blob/main/LICENSE">
    <img src="https://img.shields.io/npm/l/axios-mapper" alt="License">
  </a>
</p>

axios-mapper can make  ts  project easy to transform model  🚀🚀 and   Prevent duplicate network requests

## Install
```shell
  npm install  axios-mapper
```
or

```shell
  yarn add axios-mapper
```
## Features

- [x] axios easy to request and  return data  auto  transform  model
- [x] prevent duplicate network requests


## Usage
&nbsp;
1、base  setting
```js
 import HttpClient,{HttpClientConfig} from "../src/index";
 const config:HttpClientConfig = {
    baseURL:'http://www.httpbin.org',
    headers:{
        token:'your token'
    }    
 }
const https = new HttpClient(config)
export default https
```

2 、use  tool  get model

vscode extension : [json2ts](https://marketplace.visualstudio.com/items?itemName=GregorBiswanger.json2ts)
web：[http://json2ts.com](http://json2ts.com)

```javascript
// {
//     "slideshow": {
//       "author": "Yours Truly",
//       "date": "date of publication",
//       "slides": [
//         {
//           "title": "Wake up to WonderWidgets!",
//           "type": "all"
//         },
//         {
//           "items": [
//             "Why <em>WonderWidgets</em> are great",
//             "Who <em>buys</em> WonderWidgets"
//           ],
//           "title": "Overview",
//           "type": "all"
//         }
//       ],
//       "title": "Sample Slide Show"
//     }
//   }

export interface Slide {
	title: string;
	type: string;
}

export interface Slideshow {
	author: string;
	date: string;
	slides: Slide[];
	title: string;
}

export interface RootObject {
	slideshow: Slideshow;
}
```

3、request

```javascript
import  https  from "./http";
import { RootObject } from "./model";

https.request<RootObject>('/json').then((res)=>{
    console.log(res?.slideshow);
})
```


### Dependency

 - axios
 - qs

### Scripts

use  [tsdx](https://github.com/formium/tsdx)  to publish

## License
axios-mapper: Axios is open-sourced software licensed under the [MIT license](./LICENSE).


CHANGELOG.md
## [0.5.7](https://github.com/RainManGO/axios-mapper/compare/0.5.2...0.5.7) (2021-09-02)


### Bug Fixes

* **index.ts:** 暴露defaultConfig ([d60b0b3](https://github.com/RainManGO/axios-mapper/commit/d60b0b3263e569db6d32535fa5d7903123fb0d0f))





test/http.ts
/*
 * @Description:  httpClient
 * @Author: ZY
 * @Date: 2020-12-11 13:42:27
 * @LastEditors: ZY
 * @LastEditTime: 2021-09-02 16:54:25
 */

import HttpClient, { HttpClientConfig } from '../src/index';

const config: HttpClientConfig = {
  baseURL: 'http://flash-admin.enilu.cn/prod-api',
  headers: {
    token: '',
  },
};

const https = new HttpClient(config);

// Add a request interceptor
https.httpClient.interceptors.request.use(
  function(config) {
    // Do something before request is sent
    return config;
  },
  function(error) {
    // Do something with request error
    return Promise.reject(error);
  }
);

// Add a response interceptor
https.httpClient.interceptors.response.use(
  function(response) {
    // Do something with response data
    return response;
  },
  function(error) {
    // Do something with response error
    return Promise.reject(error);
  }
);

export default https;


test/blah.test.ts
/*
 * @Description:
 * @Author: ZY
 * @Date: 2021-02-02 15:32:28
 * @LastEditors: ZY
 * @LastEditTime: 2021-02-02 17:23:31
 */
import https from './http';
import { Method, ContentType } from '../src/type';
import { RootObject } from './model';
import { LoginRoot } from './loginModel';

// https.request<RootObject>('/json').then((res)=>{
//     console.log(res?.slideshow);
// })

const userInfo = {
  username: 'admin',
  password: 'admin',
};

describe('post', () => {
  it('works', () => {
    https
      .request<LoginRoot>(
        'account/login',
        Method.POST,
        userInfo,
        ContentType.form
      )
      .then(res => {
        console.log(res);
      });
  });
});

describe('get', () => {
  it('works', () => {
    https.request<RootObject>('/json').then(res => {
      console.log(res?.slideshow);
    });
  });
});


test/loginModel.ts
export interface LoginModel {
  token: string;
}

export interface LoginRoot {
  code: number;
  msg: string;
  data: LoginModel;
  success: boolean;
}


src/index.ts
/*
 * @Description: network request json transform model (网络请求直接返回model)
 * @Author: ZY
 * @Date: 2020-12-11 09:40:18
 * @LastEditors: ZY
 * @LastEditTime: 2021-09-08 10:43:52
 */
import { RequestParams, Method, ContentType } from './type';
import axios, { AxiosInstance, AxiosRequestConfig } from 'axios';
import { Convert } from './json2Model';
import DuplicateRequest from './duplicate';
import qs from 'qs';

export * from './type';

export interface HttpClientConfig extends AxiosRequestConfig {
  defaultParams?: RequestParams;
  //是否开启打印
  log?: boolean;
  //是否开启检查快速点击事件
  checkQuickClick?: boolean;
  //click interval (点击间隔时间)
  clickInterval?: number;
}

export default class HttpClient {
  public defaultConfig: HttpClientConfig;
  public httpClient: AxiosInstance;

  constructor(options: HttpClientConfig = {}) {
    this.httpClient = axios.create(options);
    this.defaultConfig = options;
  }

  /**
   * @description: 封装请求类
   * @param {Method} method 请求方式
   * @param {APIPath} path 请求路径
   * @param {RequestParams} params 参数
   * @param {ContentType} contentType http配置
   * @param {RequestOptions} optionsSource
   * @return {*}
   */
  async request<T>(
    path: string = '',
    method: Method = Method.GET,
    params?: RequestParams,
    contentType: ContentType = ContentType.json,
    optionsSource?: HttpClientConfig
  ) {
    const options: HttpClientConfig = Object.assign(
      {},
      this.defaultConfig,
      optionsSource
    );
    const { headers, clickInterval } = options;
    headers['content-type'] = contentType;
    const allParams = Object.assign(
      {},
      this.defaultConfig.defaultParams,
      params
    );

    const requestConfig: HttpClientConfig = {
      url: `${path}`,
      method,
      headers,
    };

    if (
      options.checkQuickClick === true &&
      DuplicateRequest.hashUrlAndParams(
        requestConfig.url ?? '',
        method,
        allParams,
        clickInterval
      )
    ) {
      options.log && console.log('click quick');
      return null;
    }

    if (method === Method.GET) {
      requestConfig.params = allParams;
    } else {
      if (contentType === ContentType.form) {
        requestConfig.data = qs.stringify(allParams);
      } else {
        requestConfig.data = JSON.stringify(allParams);
      }
    }
    return this.httpClient
      .request(requestConfig)
      .then(res => {
        options.log && console.log(res);
        const data: string = JSON.stringify(res.data);
        if (res.status >= 200 && res.status < 300) {
          return Convert.jsonToModel(data) as T;
        } else {
          return Promise.reject(data);
        }
      })
      .catch(async error => {
        return Promise.reject(error);
      });
  }
}


src/duplicate.ts
/*
 * @Description: Prevent duplicate requests（防止不必要的请求）
 * @Author: ZY
 * @Date: 2020-12-11 09:56:27
 * @LastEditors: ZY
 * @LastEditTime: 2020-12-14 11:15:02
 */
import qs from 'qs';
import { RequestParams, Method } from './type';

export default class DuplicateRequest {
  private static _instance: DuplicateRequest;
  private constructor() {}
  static getInstance() {
    if (!this._instance) {
      this._instance = new DuplicateRequest();
    }
    return this._instance;
  }

  static lastRequestTimeStamp = 0;
  static lastIdentityCode = '';

  /**
   * @description: 验证是否是相同的url点击过快
   * @param {string} url
   * @param {Method} method
   * @param {RequestParams} params
   * @return {*}
   */
  static hashUrlAndParams(
    url: string,
    method: Method,
    params: RequestParams,
    interval: number = 1000
  ) {
    const identityCode = [url, method, qs.stringify(params)].join('&');
    const currentTime = Date.parse(new Date().toString());
    if (
      currentTime - DuplicateRequest.lastRequestTimeStamp < interval &&
      DuplicateRequest.lastIdentityCode === identityCode
    ) {
      return true;
    } else {
      DuplicateRequest.lastRequestTimeStamp = currentTime;
      DuplicateRequest.lastIdentityCode = identityCode;
      return false;
    }
  }
}


example/model.ts
/*
 * @Description:
 * @Author: ZY
 * @Date: 2020-12-11 13:52:11
 * @LastEditors: ZY
 * @LastEditTime: 2020-12-14 15:28:49
 */

// {
//     "slideshow": {
//       "author": "Yours Truly",
//       "date": "date of publication",
//       "slides": [
//         {
//           "title": "Wake up to WonderWidgets!",
//           "type": "all"
//         },
//         {
//           "items": [
//             "Why <em>WonderWidgets</em> are great",
//             "Who <em>buys</em> WonderWidgets"
//           ],
//           "title": "Overview",
//           "type": "all"
//         }
//       ],
//       "title": "Sample Slide Show"
//     }
//   }

export interface Slide {
  title: string;
  type: string;
}

export interface Slideshow {
  author: string;
  date: string;
  slides: Slide[];
  title: string;
}

export interface RootObject {
  slideshow: Slideshow;
}

export interface Arg {}

export interface File {}

export interface Form {}

export interface Header {
  accept: string;
  'accept-Encoding': string;
  'accept-Language': string;
  'content-Length': string;
  host: string;
  origin: string;
  referer: string;
  'user-Agent': string;
  'x-Amzn-Trace-Id': string;
}

export interface RootObjectPost {
  args: Arg;
  data: string;
  files: File;
  form: Form;
  headers: Header;
  json?: any;
  method: string;
  origin: string;
  url: string;
}
